substrate/frame/opf/src/lib.rs

```diff
@@ -0,0 +1,333 @@
+
+		/// The users voting period ended. Reward calculation will start.
+		VoteActionLocked { when: BlockNumberFor<T>, round_number: u32 },
+
+		/// The voting round ended
+		VotingRoundEnded { when: BlockNumberFor<T>, round_number: u32 },
+
+		/// User's funds unlocked
+		FundsUnlocked { when: BlockNumberFor<T>, amount: BalanceOf<T>, project_id: AccountIdOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// This account is not connected to any WhiteListed Project.
+		NotWhitelistedProject,
+
+		/// There are no whitelisted project
+		NoWhitelistedProject,
+
+		/// The voting action failed.
+		VoteFailed,
+
+		/// No such voting data
+		NoVoteData,
+
+		/// An invalid result  was returned
+		InvalidResult,
+
+		/// Maximum number of projects submission for distribution as been reached
+		MaximumProjectsNumber,
+
+		/// This voting round does not exists
+		NoRoundFound,
+
+		/// Voting period closed for this round
+		VotePeriodClosed,
+
+		/// Not enough funds to vote, you need to decrease your stake
+		NotEnoughFunds,
+
+		/// The reward calculation failed due to an internal error
+		FailedRewardCalculation,
+
+		/// Voting round is over
+		VotingRoundOver,
+
+		/// User's funds still cannot be unlocked
+		FundsUnlockNotPermitted,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Weight: see `begin_block`
+		fn on_idle(n: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			Self::on_idle_function(n, remaining_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// OPF voting logic
+		///
+		/// ## Dispatch Origin
+		///
+		/// Must be signed
+		///
+		/// ## Details
+		///
+		/// This extrinsic allows users to [vote for/nominate] a whitelisted project using their funds.
+		/// As a first implementation, the `conviction` parameter was not included for simplicity, but /// should be in the next iteration of the pallet.
+		/// The amount defined by the user is locked and released only when the project reward is /// sent for distribution, or when the project is not dimmed fundable.
+		/// Users can edit an existing vote within the vote-casting period.
+		/// Then, during the vote-locked period, rewards are calculated based on the total user amount
+		/// attributed to each project by the user’s votes.
+		///
+		/// ### Parameters
+		/// - `project_account`: The account that will receive the reward.
+		/// - `amount`: Amount that will be locked in user’s balance to nominate a project.
+		/// - `is_fund`: Parameter that defines if user’s vote is in favor (*true*), or against (*false*)
+		/// the project funding.
+
+		/// ### Errors
+		/// - [`Error::<T>::NotEnoughFunds`]: The user does not have enough balance to cast a vote
+		///  
+		/// ## Events
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::vote(T::MaxWhitelistedProjects::get()))]
+		pub fn vote(
+			origin: OriginFor<T>,
+			project_account: ProjectId<T>,
+			amount: BalanceOf<T>,
```

Comment:
```suggestion
			#[pallet::compact] amount: BalanceOf<T>,
```

generally compact encoding should be used for all Balance type to avodi some encoded bytes